contracts/NodeVersionBeacon.cdc

```diff
@@ -280,6 +285,13 @@ pub contract NodeVersionBeacon {
             
             emit NodeVersionBoundaryFreezePeriodChanged(freezePeriod: newFreezePeriod)
         }
+
+        /// Adds a pending protocol state version upgrade, which will be emitted
+        /// as a service event during the next heartbeat.
+        pub fun setPendingProtocolStateVersionUpgrade(newProtocolVersion: UInt64, activeView: UInt64) {
+            let pendingUpgrade = NodeVersionBeacon.ViewActivatedVersion(version: newProtocolVersion, activeView: activeView)
+            NodeVersionBeacon.storePendingProtocolStateVersionUpgrade(upgrade: pendingUpgrade)
```

Comment:
Just to confirm my understanding:

The reason we don't just emit the event here is that the service events are only observed from the service transaction.

Reply:
Exactly. The idea was to be as restrictive as possible for what qualifies as a service event, because they are so security-sensitive, at the cost of convenience in this kind of case.